### PR TITLE
Unit Testing Production ML Code - Code Overview

### PR DESCRIPTION
**Pull request by ChristopherGS**

### CODE REVIEW

File: packages/gradient_boosting_model/gradient_boosting_model/datasets/houseprice.csv [hunk 1]
```diff
@@ -0,0 +1,1461 @@
+Id,MSSubClass,MSZoning,LotFrontage,LotArea,Street,Alley,LotShape,LandContour,Utilities,LotConfig,LandSlope,Neighborhood,Condition1,Condition2,BldgType,HouseStyle,OverallQual,OverallCond,YearBuilt,YearRemodAdd,RoofStyle,RoofMatl,Exterior1st,Exterior2nd,MasVnrType,MasVnrArea,ExterQual,ExterCond,Foundation,BsmtQual,BsmtCond,BsmtExposure,BsmtFinType1,BsmtFinSF1,BsmtFinType2,BsmtFinSF2,BsmtUnfSF,TotalBsmtSF,Heating,HeatingQC,CentralAir,Electrical,1stFlrSF,2ndFlrSF,LowQualFinSF,GrLivArea,BsmtFullBath,BsmtHalfBath,FullBath,HalfBath,BedroomAbvGr,KitchenAbvGr,KitchenQual,TotRmsAbvGrd,Functional,Fireplaces,FireplaceQu,GarageType,GarageYrBlt,GarageFinish,GarageCars,GarageArea,GarageQual,GarageCond,PavedDrive,WoodDeckSF,OpenPorchSF,EnclosedPorch,3SsnPorch,ScreenPorch,PoolArea,PoolQC,Fence,MiscFeature,MiscVal,MoSold,YrSold,SaleType,SaleCondition,SalePrice
```
Comment (solegalli) [inline packages/gradient_boosting_model/gradient_boosting_model/datasets/houseprice.csv]: Hi Chris. We are not allowed to re-host the data set. That is a violation of copyright. We do have permission from Prof DeCock to use it, on the condition we give him a copy of the product we built using his data set. He has given us a link to the original data. We are however using the data set from Kaggle. We do have permission from Kaggle to direct students to download the data from their site. But re-hosting is not allowed. Can we re do this so that they download the data set themselves, and therefore can accept the T&C from Kaggle when they do so? Or alternatively, use the Kaggle API for downloading. We need to remove the dataset from the commit history as well. Sorry about this. Cheers. Sole @ChristopherGS

File: packages/gradient_boosting_model/gradient_boosting_model/pipeline.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
+from sklearn.impute import SimpleImputer
+from sklearn.preprocessing import OrdinalEncoder
+from sklearn.ensemble import GradientBoostingRegressor
+from sklearn.pipeline import Pipeline
+from feature_engine.categorical_encoders import RareLabelCategoricalEncoder
```
Comment (solegalli) [inline packages/gradient_boosting_model/gradient_boosting_model/pipeline.py:5]: Make sure we are using latest version of Feature-engine. Shouldn't affect the rarelabelencoder tho.

File: packages/gradient_boosting_model/gradient_boosting_model/processing/features.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
+import numpy as np
+from sklearn.base import BaseEstimator, TransformerMixin
+
+from gradient_boosting_model.processing.errors import InvalidModelInputError
+
+
+class LogTransformer(BaseEstimator, TransformerMixin):
```
Comment (solegalli) [inline packages/gradient_boosting_model/gradient_boosting_model/processing/features.py]: We shouldn't have to apply a log transform to the variables if using gradient boosting machines. This was for the linear regression. Why is this here?

File: packages/gradient_boosting_model/gradient_boosting_model/processing/preprocessors.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
+import pandas as pd
+from sklearn.base import BaseEstimator, TransformerMixin
+
+
+class SklearnTransformerWrapper(BaseEstimator, TransformerMixin):
+    """Categorical data missing value imputer."""
```
Comment (solegalli) [inline packages/gradient_boosting_model/gradient_boosting_model/processing/preprocessors.py]: The transformer wrapper is a wrapper around sklearn pre processing transformers, so that we can select variables. We need to replace the text in the function:
" Wrapper for Scikit-learn pre-processing transformers,like the SimpleImputer() or OrdinalEncoder(), to allow the use of the transformer on a selected group of variables"

File: packages/gradient_boosting_model/gradient_boosting_model/processing/preprocessors.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+        if not isinstance(variables, list):
+            self.variables = [variables]
+        else:
+            self.variables = variables
+
+        self.transformer = transformer
+
+    def fit(self, X: pd.DataFrame, y: pd.Series = None):
+        """Fit statement to accomodate the sklearn pipeline."""
```
Comment (solegalli) [inline packages/gradient_boosting_model/gradient_boosting_model/processing/preprocessors.py]: Here the fit() mode is doing something, so we do not need to add that comment. That is for then fit() returns self.
Here I would say:
"Fit function allows Scikit-learn transformers to learn the required parameters from the train set"

File: packages/gradient_boosting_model/gradient_boosting_model/processing/preprocessors.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+
+    def transform(self, X: pd.DataFrame) -> pd.DataFrame:
+        """Apply the transforms to the dataframe."""
+        X = X.copy()
+        X[self.variables] = self.transformer.transform(X[self.variables])
+        return X
+
+
+class TemporalVariableEstimator(BaseEstimator, TransformerMixin):
+    """Temporal variable calculator."""
```
Comment (solegalli) [inline packages/gradient_boosting_model/gradient_boosting_model/processing/preprocessors.py]: Replace by:
"Calculates the time difference between 2 temporal variables"

File: packages/gradient_boosting_model/gradient_boosting_model/processing/preprocessors.py [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+
+        return X
+
+
+class DropUnecessaryFeatures(BaseEstimator, TransformerMixin):
+    def __init__(self, variables_to_drop=None):
+        self.variables = variables_to_drop
+
+    def fit(self, X, y=None):
+        return self
```
Comment (solegalli) [inline packages/gradient_boosting_model/gradient_boosting_model/processing/preprocessors.py:74]: Here I would add:
"fit() method is necessary to accomodate the Scikit-learn pipeline functionality"

File: packages/gradient_boosting_model/gradient_boosting_model/processing/preprocessors.py [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+
+    def transform(self, X):
+        # encode labels
```
Comment (solegalli) [inline packages/gradient_boosting_model/gradient_boosting_model/processing/preprocessors.py]: Comment here is wrong, should be:
"drop unnecesary / unused features from the data set"

File: packages/gradient_boosting_model/requirements.txt [hunk 9]
```diff
@@ -0,0 +1,14 @@
+# production requirements
+numpy>=1.17.3,<1.18.0
+scikit-learn>=0.21.3,<0.22.0
+pandas>=0.25.3,<0.26.0
+feature_engine==0.3.0
```
Comment (solegalli) [inline packages/gradient_boosting_model/requirements.txt:5]: ok, this is the right version

File: packages/gradient_boosting_model/setup.py [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+import io
+import os
+
+from setuptools import find_packages, setup
+from gradient_boosting_model import config, PACKAGE_ROOT
+
+
+# Package meta-data.
+NAME = config['package_name'].data
+DESCRIPTION = "Train and deploy LARS Lasso regression model."
```
Comment (solegalli) [inline packages/gradient_boosting_model/setup.py]: what is LARS? should be removed I think

Comment (solegalli) [reply packages/gradient_boosting_model/setup.py]: Also, is this for the logit, or the gbm? I am confused now, are we building both models? one?

File: packages/gradient_boosting_model/gradient_boosting_model/datasets/data_description.txt [hunk 11]
```diff
@@ -0,0 +1,523 @@
+MSSubClass: Identifies the type of dwelling involved in the sale.
```
Comment (solegalli) [inline packages/gradient_boosting_model/gradient_boosting_model/datasets/data_description.txt]: I am also uncertain as to wheter I would include this in the repo. This information is together with the data set and can be viewed at the right source. See my comment on the data set re-hosting. I would remove from the repo. Copyright potential issues.

**Review by solegalli — CHANGES_REQUESTED** [review]
@ChristopherGS , very clear the entire code, well commented, clear variable names, very easy to follow, for someone like me :p 
Students may require a bit more of handholding. 
I added a few comments here and there, nothing major, apart from the copyright issue with the data set re-hosting. I would add a lecture / article at the beginning with intructions to download the data set, as we do in other courses. And we need to remove the data set from the commit history as well.
Hope this helps. Cheers
Sole

File: packages/gradient_boosting_model/gradient_boosting_model/datasets/houseprice.csv (hunk 1 above)
Comment (ChristopherGS) [reply packages/gradient_boosting_model/gradient_boosting_model/datasets/houseprice.csv]: Done

File: packages/gradient_boosting_model/gradient_boosting_model/pipeline.py (hunk 2 above)
Comment (ChristopherGS) [reply packages/gradient_boosting_model/gradient_boosting_model/pipeline.py:5]: We're using v 0.3.0 which I believe is the latest

File: packages/gradient_boosting_model/gradient_boosting_model/processing/features.py (hunk 3 above)
Comment (ChristopherGS) [reply packages/gradient_boosting_model/gradient_boosting_model/processing/features.py]: removed

File: packages/gradient_boosting_model/gradient_boosting_model/processing/preprocessors.py (hunk 4 above)
Comment (ChristopherGS) [reply packages/gradient_boosting_model/gradient_boosting_model/processing/preprocessors.py]: updated

File: packages/gradient_boosting_model/gradient_boosting_model/processing/preprocessors.py (hunk 5 above)
Comment (ChristopherGS) [reply packages/gradient_boosting_model/gradient_boosting_model/processing/preprocessors.py]: updated

File: packages/gradient_boosting_model/gradient_boosting_model/processing/preprocessors.py (hunk 6 above)
Comment (ChristopherGS) [reply packages/gradient_boosting_model/gradient_boosting_model/processing/preprocessors.py]: updated

File: packages/gradient_boosting_model/gradient_boosting_model/processing/preprocessors.py (hunk 7 above)
Comment (ChristopherGS) [reply packages/gradient_boosting_model/gradient_boosting_model/processing/preprocessors.py:74]: updated

File: packages/gradient_boosting_model/gradient_boosting_model/processing/preprocessors.py (hunk 8 above)
Comment (ChristopherGS) [reply packages/gradient_boosting_model/gradient_boosting_model/processing/preprocessors.py]: fixed

File: packages/gradient_boosting_model/setup.py (hunk 10 above)
Comment (ChristopherGS) [reply packages/gradient_boosting_model/setup.py]: gbm. Fixed

File: packages/gradient_boosting_model/gradient_boosting_model/datasets/data_description.txt (hunk 11 above)
Comment (ChristopherGS) [reply packages/gradient_boosting_model/gradient_boosting_model/datasets/data_description.txt]: removed
